# Scraping Historical Starting Lineups 

#### https://www.basketball-reference.com/teams/{}/{}_start.html

In [2]:
import pandas as pd
from urllib.request import urlopen
from bs4 import BeautifulSoup
import sys 
sys.path.append('../Scripts')
import starting_lineup_functions as slp

../Scripts\starting_lineup_functions.py:11: UserWarning: DataFrame columns are not unique, some columns will be omitted.
  PLAYER_DICT = players.T.to_dict('list')


In [212]:
team_names = slp.PAST_DICT
team_name_list = []

for values in team_names.values():
    team_name_list += values

In [213]:
team_name_list.remove('CHB')

In [214]:
team_name_list

['ATL',
 'BKN',
 'BOS',
 'CHA',
 'CHI',
 'CLE',
 'DAL',
 'DEN',
 'DET',
 'GSW',
 'HOU',
 'IND',
 'LAC',
 'LAL',
 'MEM',
 'MIA',
 'MIL',
 'MIN',
 'NOP',
 'NOP',
 'NYK',
 'OKC',
 'ORL',
 'PHI',
 'PHX',
 'POR',
 'SAC',
 'SAS',
 'TOR',
 'UTA',
 'WAS']

In [201]:
def starting_lineup_test(team, year):
    
    url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(team, year)
    html = urlopen(url)
    soup = BeautifulSoup(html)
    tr_rows = soup.find_all('tr')[1:51]
    lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_df = pd.DataFrame(lineup_data, columns=header)
    lineup_data_df['Team'] = team 
    lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
    
    return lineup_data_df

In [4]:
def starting_lineup():
        
    team_name_list = ['ATL','BOS','BRK', 'CHI', 'CLE', 'DAL', 'DEN', 'DET', 'GSW', 'HOU', 'IND', 'LAC', 'LAL', 'MEM', 'MIA',
                       'MIL', 'MIN', 'NYK', 'OKC', 'ORL', 'PHI', 'PHO', 'POR', 'SAC', 'SAS', 'TOR', 'UTA', 'WAS']
    
    
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_final = pd.DataFrame()
    year = [2013, 2014, 2015, 2016, 2017, 2018, 2019]
    
    for i in team_name_list:
        for j in year:  
            url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(i, j)
            html = urlopen(url)
            soup = BeautifulSoup(html)
            tr_rows = soup.find_all('tr')[1:83]
            lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
            lineup_data_df = pd.DataFrame(lineup_data, columns=header)
            lineup_data_df['Team'] = i 
            lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
            lineup_data_final = lineup_data_final.append(lineup_data_df)
    
    
    return lineup_data_final

In [5]:
starting_lineup_master = starting_lineup()
starting_lineup_master

Game Date  Box Score Team     Player1     Player2       Player3  \
0  2012-11-02  Box Score  ATL   D. Harris  A. Horford     K. Korver   
1  2012-11-04  Box Score  ATL  A. Horford   K. Korver  D. Stevenson   
2  2012-11-07  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
3  2012-11-09  Box Score  ATL  A. Horford   K. Korver      J. Smith   
4  2012-11-11  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
..        ...        ...  ...         ...         ...           ...   
77 2019-03-31  Box Score  WAS     B. Beal    T. Brown     T. Bryant   
78 2019-04-03  Box Score  WAS     B. Beal    T. Brown     T. Bryant   
79 2019-04-05  Box Score  WAS     B. Beal    T. Brown     T. Bryant   
80 2019-04-07  Box Score  WAS     B. Beal    T. Brown     T. Bryant   
81 2019-04-09  Box Score  WAS     B. Beal    T. Brown     T. Bryant   

         Player4        Player5  
0       J. Smith      J. Teague  
1      J. Teague    A. Tolliver  
2       J. Smith      J. Teague  
3   D. Stevenson      J. Teague  
4       J. Smith      J. Teague  
..           ...            ...  
77     B. Portis  T. Satoranský  
78     B. Portis  T. Satoranský  
79     B. Portis  T. Satoranský  
80     B. Portis  T. Satoranský  
81     B. Portis  T. Satoranský  

[16072 rows x 8 columns]

##### Data check: 28 teams x 82 games x 7 seasons = 16072 

#### Starting Lineup for 2020 season, separate function to account for less than 82 games played 

In [6]:
def starting_lineup_2020():
        
    team_name_list = ['ATL','BOS','BRK', 'CHI', 'CHO', 'CLE', 'DAL', 'DEN', 'DET', 'GSW', 'HOU', 'IND', 'LAC', 'LAL', 'MEM', 'MIA',
                       'MIL', 'MIN', 'NOP', 'NYK', 'OKC', 'ORL', 'PHI', 'PHO', 'POR', 'SAC', 'SAS', 'TOR', 'UTA', 'WAS']
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_final = pd.DataFrame()
    year = [2020]
    
    for i in team_name_list:
        for j in year:  
            url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(i, j)
            html = urlopen(url)
            soup = BeautifulSoup(html)
            tr_rows = soup.find_all('tr')[1:51]
            lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
            lineup_data_df = pd.DataFrame(lineup_data, columns=header)
            lineup_data_df['Team'] = i 
            lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
            lineup_data_final = lineup_data_final.append(lineup_data_df)
    
    
    return lineup_data_final

In [7]:
starting_lineup_2020 = starting_lineup_2020()
starting_lineup_2020

Game Date  Box Score Team     Player1       Player2       Player3  \
0  2019-10-24  Box Score  ATL  J. Collins     D. Hunter        A. Len   
1  2019-10-26  Box Score  ATL  J. Collins     D. Hunter        A. Len   
2  2019-10-28  Box Score  ATL  J. Collins     D. Hunter        A. Len   
3  2019-10-29  Box Score  ATL  J. Collins     D. Hunter        A. Len   
4  2019-10-31  Box Score  ATL  J. Collins    K. Huerter     D. Hunter   
..        ...        ...  ...         ...           ...           ...   
45 2020-01-28  Box Score  WAS     B. Beal      I. Bonga     T. Bryant   
46 2020-01-30  Box Score  WAS     B. Beal      I. Bonga    I. Mahinmi   
47 2020-02-01  Box Score  WAS     B. Beal      I. Bonga    I. Mahinmi   
48 2020-02-03  Box Score  WAS     B. Beal  R. Hachimura    I. Mahinmi   
49 2020-02-07  Box Score  WAS     B. Beal      I. Bonga  R. Hachimura   

       Player4     Player5  
0   C. Reddish    T. Young  
1   C. Reddish    T. Young  
2   C. Reddish    T. Young  
3   C. Reddish    T. Young  
4       A. Len  C. Reddish  
..         ...         ...  
45  I. Mahinmi   I. Thomas  
46   G. Payton   I. Thomas  
47   G. Payton   I. Thomas  
48   G. Payton   I. Thomas  
49  I. Mahinmi    I. Smith  

[1500 rows x 8 columns]

##### Data check: 30 teams x 49 games x 1 season = 1500 

In [8]:
starting_lineup_master = starting_lineup_master.append(starting_lineup_2020)

In [9]:
starting_lineup_master

Game Date  Box Score Team     Player1       Player2       Player3  \
0  2012-11-02  Box Score  ATL   D. Harris    A. Horford     K. Korver   
1  2012-11-04  Box Score  ATL  A. Horford     K. Korver  D. Stevenson   
2  2012-11-07  Box Score  ATL  A. Horford     K. Korver   Z. Pachulia   
3  2012-11-09  Box Score  ATL  A. Horford     K. Korver      J. Smith   
4  2012-11-11  Box Score  ATL  A. Horford     K. Korver   Z. Pachulia   
..        ...        ...  ...         ...           ...           ...   
45 2020-01-28  Box Score  WAS     B. Beal      I. Bonga     T. Bryant   
46 2020-01-30  Box Score  WAS     B. Beal      I. Bonga    I. Mahinmi   
47 2020-02-01  Box Score  WAS     B. Beal      I. Bonga    I. Mahinmi   
48 2020-02-03  Box Score  WAS     B. Beal  R. Hachimura    I. Mahinmi   
49 2020-02-07  Box Score  WAS     B. Beal      I. Bonga  R. Hachimura   

         Player4      Player5  
0       J. Smith    J. Teague  
1      J. Teague  A. Tolliver  
2       J. Smith    J. Teague  
3   D. Stevenson    J. Teague  
4       J. Smith    J. Teague  
..           ...          ...  
45    I. Mahinmi    I. Thomas  
46     G. Payton    I. Thomas  
47     G. Payton    I. Thomas  
48     G. Payton    I. Thomas  
49    I. Mahinmi     I. Smith  

[17572 rows x 8 columns]

##### Data check: 1500 games in 2020 + 16072 games in 2014-2019 (28 teams) = 17572 

#### separate function for Charlotte Bobcats who had a name change to Charlotte Hornets in 2015 season, CHA -> CHO

In [16]:
def starting_lineup_CHO():
        
    team_name_list = ['CHO']
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_final = pd.DataFrame()
    year = [2015,2016,2017,2018,2019]
    
    for i in team_name_list:
        for j in year:  
            url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(i, j)
            html = urlopen(url)
            soup = BeautifulSoup(html)
            tr_rows = soup.find_all('tr')[1:83]
            lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
            lineup_data_df = pd.DataFrame(lineup_data, columns=header)
            lineup_data_df['Team'] = i 
            lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
            lineup_data_final = lineup_data_final.append(lineup_data_df)
    
    
    return lineup_data_final

In [17]:
starting_lineup_CHO_2015_2019 = starting_lineup_CHO()

In [18]:
def starting_lineup_CHA():
        
    team_name_list = ['CHA']
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_final = pd.DataFrame()
    year = [2013,2014]
    
    for i in team_name_list:
        for j in year:  
            url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(i, j)
            html = urlopen(url)
            soup = BeautifulSoup(html)
            tr_rows = soup.find_all('tr')[1:83]
            lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
            lineup_data_df = pd.DataFrame(lineup_data, columns=header)
            lineup_data_df['Team'] = 'CHO'
            lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
            lineup_data_final = lineup_data_final.append(lineup_data_df)
    
    
    return lineup_data_final

In [19]:
starting_lineup_CHA_2013_2014 = starting_lineup_CHA()

### combines all seasons under 'CHO', Charlotte Hornets

In [20]:
starting_lineup_CHO_All = starting_lineup_CHA_2013_2014.append(starting_lineup_CHO_2015_2019)
starting_lineup_CHO_All

Game Date  Box Score Team     Player1            Player2  \
0  2012-11-02  Box Score  CHO  B. Haywood       G. Henderson   
1  2012-11-03  Box Score  CHO  B. Haywood       G. Henderson   
2  2012-11-07  Box Score  CHO  B. Haywood  M. Kidd-Gilchrist   
3  2012-11-09  Box Score  CHO  B. Biyombo  M. Kidd-Gilchrist   
4  2012-11-10  Box Score  CHO  B. Haywood  M. Kidd-Gilchrist   
..        ...        ...  ...         ...                ...   
77 2019-04-03  Box Score  CHO    D. Bacon         B. Biyombo   
78 2019-04-05  Box Score  CHO    D. Bacon         B. Biyombo   
79 2019-04-07  Box Score  CHO    D. Bacon           N. Batum   
80 2019-04-09  Box Score  CHO    D. Bacon           N. Batum   
81 2019-04-10  Box Score  CHO    D. Bacon           N. Batum   

              Player3            Player4    Player5  
0   M. Kidd-Gilchrist         B. Mullens  K. Walker  
1   M. Kidd-Gilchrist         B. Mullens  K. Walker  
2          B. Mullens          J. Taylor  K. Walker  
3          B. Mullens          J. Taylor  K. Walker  
4          B. Mullens          J. Taylor  K. Walker  
..                ...                ...        ...  
77         M. Bridges  M. Kidd-Gilchrist  K. Walker  
78         M. Bridges  M. Kidd-Gilchrist  K. Walker  
79         B. Biyombo         M. Bridges  K. Walker  
80         B. Biyombo         M. Bridges  K. Walker  
81         B. Biyombo         M. Bridges  K. Walker  

[574 rows x 8 columns]

#### Data Check: 1 team x 7 seasons x 82 games = 574

#### separate function for New Orleans Pelicans who had a name change in 2015 season, NOH -> NOP

In [22]:
#New Orleans Pelicans name change in 2014 season, from NOH -> NOP
def starting_lineup_NOP():
        
    team_name_list = ['NOP']
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_final = pd.DataFrame()
    year = [2014, 2015,2016,2017,2018,2019]
    
    for i in team_name_list:
        for j in year:  
            url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(i, j)
            html = urlopen(url)
            soup = BeautifulSoup(html)
            tr_rows = soup.find_all('tr')[1:83]
            lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
            lineup_data_df = pd.DataFrame(lineup_data, columns=header)
            lineup_data_df['Team'] = i 
            lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
            lineup_data_final = lineup_data_final.append(lineup_data_df)
    
    
    return lineup_data_final

In [23]:
starting_lineup_NOP_2014_2019 = starting_lineup_NOP()

In [24]:
def starting_lineup_NOH():
        
    team_name_list = ['NOH']
    header = ['Game Date', 'Box Score', 'Team', 'Player1', 'Player2', 'Player3', 'Player4', 'Player5']
    lineup_data_final = pd.DataFrame()
    year = [2013]
    
    for i in team_name_list:
        for j in year:  
            url = 'https://www.basketball-reference.com/teams/{}/{}_start.html'.format(i, j)
            html = urlopen(url)
            soup = BeautifulSoup(html)
            tr_rows = soup.find_all('tr')[1:83]
            lineup_data = [[a.get_text() for a in tr_rows[i].find_all('a')] for i in range(len(tr_rows))]
            lineup_data_df = pd.DataFrame(lineup_data, columns=header)
            lineup_data_df['Team'] = 'NOP'
            lineup_data_df['Game Date'] = pd.to_datetime(lineup_data_df['Game Date'])
            lineup_data_final = lineup_data_final.append(lineup_data_df)
    
    
    return lineup_data_final

In [25]:
starting_lineup_NOH_2013 = starting_lineup_NOH()

In [26]:
starting_lineup_NOP_All = starting_lineup_NOH_2013.append(starting_lineup_NOP_2014_2019)
starting_lineup_NOP_All

Game Date  Box Score Team   Player1      Player2    Player3      Player4  \
0  2012-10-31  Box Score  NOP  A. Aminu     A. Davis   R. Lopez    A. Rivers   
1  2012-11-02  Box Score  NOP  A. Aminu     A. Davis   R. Lopez    A. Rivers   
2  2012-11-03  Box Score  NOP  A. Aminu  R. Anderson   R. Lopez    A. Rivers   
3  2012-11-07  Box Score  NOP  A. Aminu  R. Anderson   R. Lopez    D. Miller   
4  2012-11-09  Box Score  NOP  A. Aminu     A. Davis   R. Lopez    A. Rivers   
..        ...        ...  ...       ...          ...        ...          ...   
77 2019-03-31  Box Score  NOP  I. Clark    J. Okafor  E. Payton    J. Randle   
78 2019-04-03  Box Score  NOP  I. Clark    J. Okafor  E. Payton    J. Randle   
79 2019-04-05  Box Score  NOP  I. Clark    J. Okafor  E. Payton    J. Randle   
80 2019-04-07  Box Score  NOP  I. Clark    J. Okafor  E. Payton  K. Williams   
81 2019-04-09  Box Score  NOP  I. Clark    J. Okafor  E. Payton  K. Williams   

        Player5  
0    G. Vásquez  
1    G. Vásquez  
2    G. Vásquez  
3    G. Vásquez  
4    G. Vásquez  
..          ...  
77  K. Williams  
78  K. Williams  
79  K. Williams  
80      C. Wood  
81      C. Wood  

[574 rows x 8 columns]

#### Data Check 1 team x 7 seasons x 82 games  = 574

In [27]:
starting_lineup_master = starting_lineup_master.append(starting_lineup_CHO_All)
starting_lineup_master

Game Date  Box Score Team     Player1     Player2       Player3  \
0  2012-11-02  Box Score  ATL   D. Harris  A. Horford     K. Korver   
1  2012-11-04  Box Score  ATL  A. Horford   K. Korver  D. Stevenson   
2  2012-11-07  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
3  2012-11-09  Box Score  ATL  A. Horford   K. Korver      J. Smith   
4  2012-11-11  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
..        ...        ...  ...         ...         ...           ...   
77 2019-04-03  Box Score  CHO    D. Bacon  B. Biyombo    M. Bridges   
78 2019-04-05  Box Score  CHO    D. Bacon  B. Biyombo    M. Bridges   
79 2019-04-07  Box Score  CHO    D. Bacon    N. Batum    B. Biyombo   
80 2019-04-09  Box Score  CHO    D. Bacon    N. Batum    B. Biyombo   
81 2019-04-10  Box Score  CHO    D. Bacon    N. Batum    B. Biyombo   

              Player4      Player5  
0            J. Smith    J. Teague  
1           J. Teague  A. Tolliver  
2            J. Smith    J. Teague  
3        D. Stevenson    J. Teague  
4            J. Smith    J. Teague  
..                ...          ...  
77  M. Kidd-Gilchrist    K. Walker  
78  M. Kidd-Gilchrist    K. Walker  
79         M. Bridges    K. Walker  
80         M. Bridges    K. Walker  
81         M. Bridges    K. Walker  

[18146 rows x 8 columns]

In [28]:
starting_lineup_final_2013_2020_ALL = starting_lineup_master.append(starting_lineup_NOP_All)
starting_lineup_final_2013_2020_ALL

Game Date  Box Score Team     Player1     Player2       Player3  \
0  2012-11-02  Box Score  ATL   D. Harris  A. Horford     K. Korver   
1  2012-11-04  Box Score  ATL  A. Horford   K. Korver  D. Stevenson   
2  2012-11-07  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
3  2012-11-09  Box Score  ATL  A. Horford   K. Korver      J. Smith   
4  2012-11-11  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
..        ...        ...  ...         ...         ...           ...   
77 2019-03-31  Box Score  NOP    I. Clark   J. Okafor     E. Payton   
78 2019-04-03  Box Score  NOP    I. Clark   J. Okafor     E. Payton   
79 2019-04-05  Box Score  NOP    I. Clark   J. Okafor     E. Payton   
80 2019-04-07  Box Score  NOP    I. Clark   J. Okafor     E. Payton   
81 2019-04-09  Box Score  NOP    I. Clark   J. Okafor     E. Payton   

         Player4      Player5  
0       J. Smith    J. Teague  
1      J. Teague  A. Tolliver  
2       J. Smith    J. Teague  
3   D. Stevenson    J. Teague  
4       J. Smith    J. Teague  
..           ...          ...  
77     J. Randle  K. Williams  
78     J. Randle  K. Williams  
79     J. Randle  K. Williams  
80   K. Williams      C. Wood  
81   K. Williams      C. Wood  

[18720 rows x 8 columns]

#### Final Data Check: 30 teams x 7 seasons x 82 games + (30 teams x 50 games)  = 18720

In [29]:

starting_lineup_final_2013_2020_ALL.sort_values(by=['Team', 'Game Date'])

Game Date  Box Score Team     Player1       Player2       Player3  \
0  2012-11-02  Box Score  ATL   D. Harris    A. Horford     K. Korver   
1  2012-11-04  Box Score  ATL  A. Horford     K. Korver  D. Stevenson   
2  2012-11-07  Box Score  ATL  A. Horford     K. Korver   Z. Pachulia   
3  2012-11-09  Box Score  ATL  A. Horford     K. Korver      J. Smith   
4  2012-11-11  Box Score  ATL  A. Horford     K. Korver   Z. Pachulia   
..        ...        ...  ...         ...           ...           ...   
45 2020-01-28  Box Score  WAS     B. Beal      I. Bonga     T. Bryant   
46 2020-01-30  Box Score  WAS     B. Beal      I. Bonga    I. Mahinmi   
47 2020-02-01  Box Score  WAS     B. Beal      I. Bonga    I. Mahinmi   
48 2020-02-03  Box Score  WAS     B. Beal  R. Hachimura    I. Mahinmi   
49 2020-02-07  Box Score  WAS     B. Beal      I. Bonga  R. Hachimura   

         Player4      Player5  
0       J. Smith    J. Teague  
1      J. Teague  A. Tolliver  
2       J. Smith    J. Teague  
3   D. Stevenson    J. Teague  
4       J. Smith    J. Teague  
..           ...          ...  
45    I. Mahinmi    I. Thomas  
46     G. Payton    I. Thomas  
47     G. Payton    I. Thomas  
48     G. Payton    I. Thomas  
49    I. Mahinmi     I. Smith  

[18720 rows x 8 columns]

In [ ]:
starting_lineup_final_2013_2020_ALL.to_csv('../Resources/starting_lineup_2013_2020.csv')

In [308]:
starting_lineup_final_2013_2020_ALL.head()

Game Date  Box Score Team     Player1     Player2       Player3  \
0 2012-11-02  Box Score  ATL   D. Harris  A. Horford     K. Korver   
1 2012-11-04  Box Score  ATL  A. Horford   K. Korver  D. Stevenson   
2 2012-11-07  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   
3 2012-11-09  Box Score  ATL  A. Horford   K. Korver      J. Smith   
4 2012-11-11  Box Score  ATL  A. Horford   K. Korver   Z. Pachulia   

        Player4      Player5  
0      J. Smith    J. Teague  
1     J. Teague  A. Tolliver  
2      J. Smith    J. Teague  
3  D. Stevenson    J. Teague  
4      J. Smith    J. Teague

In [30]:
#In 2013 two teams only played 81 games
starting_lineup_final_2013_2020_ALL.isnull().sum()

Game Date    2
Box Score    2
Team         0
Player1      2
Player2      2
Player3      2
Player4      2
Player5      2
dtype: int64